app/services/sso/openid_connect.rb

```diff
@@ -0,0 +1,55 @@
+module SSO
+  class OpenidConnect < Base
+    delegate :session, :to => :controller
+    attr_reader :current_user
+
+    def available?
+      controller.api_request? && bearer_token_set?
+    end
+
+    def authenticate!
+      payload = jwt_token.decode
+      user = find_or_create_user_from_jwt(payload)
+      @current_user = user
+      update_session(payload)
+      user&.login
+    rescue JWT::DecodeError
+      Rails.logger.error "JWT SSO: Failed to decode JWT."
+      nil
+    end
+
+    def authenticated?
+      self.user = User.current.presence || authenticate!
+    end
+
+    private
+
+    def jwt_token
+      @jwt_token ||= jwt_token_from_request
+    end
+
+    def jwt_token_from_request
+      token = request.authorization.split(' ')[1]
+      OidcJwtToken.new(token)
+    end
+
+    def bearer_token_set?
+      request.authorization.present? && request.authorization.start_with?('Bearer')
+    end
+
+    def update_session(payload)
+      session[:sso_method] = self.class.to_s
+      session[:expires_at] = payload["exp"]
+    end
+
+    def find_or_create_user_from_jwt(payload)
+      User.find_or_create_external_user(
+        { login: payload['preferred_username'],
```

Comment:
I have used `preferred_username`,  `email`, `given_name` and `family_name` as they are listed as the standard claims for openid-connect  - https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims.

In the case of Azure AD, `upn` needs to be mapped to `mail` in keycloak with the help of something called as `mappers` in keycloak. Other IDP's also have a machanism like auth0 has [data normalization ](https://auth0.com/docs/users/concepts/overview-user-profile#data-normalization)instead of mappers.